### ISSUE TITLE
defc error with shadow-cljs. Makes Rum unusable for me... :(

### ISSUE THREAD

**Issue by stephenBDT** [tonsky/rum#233]
```
   5 | (rum/defc hello-world [g]
   6 |   [:div
   7 |    [:h2 g]
---------------^----------------------------------------------------------------
Use of undeclared Var shadow-rum.core/g
```
This happens when shadow-cljs reloads my file (remounts the `hello-world` component)
It is not happening with repacing this rum dependency with `[ua.modnakasta/rum "0.11.0-2"]`

Steps to reproduce:
  - run `lein new shadow-cljs projectname +rum && cd projectname && yarn install && yarn shadow-cljs watch app`
  - visit `localhost:8700` and see it working
  - edit `src/projectname/core.cljs` so that `hello-world` component gets an additional argument and displays it. (I removed the display of the atom state)
  - call the `hello-world` component with some string and save the file
  - see it correctly updating in the browser
  - stop the shadow-cljs watch process
  - change the dependency in `shadow-cljs,edn` from `[ua.modnakasta/rum "0.11.0-2"]` to `[rum "0.12.4"]`
  - again run `yarn shadow-cljs watch app`, visit in the browser (maybe you already see the error
  - if not, modify `src/projectname/core.cljs` and save the file

**Comment by roman01la** [comment]
@stephenBDT Is it reproducible without shadow?

**Comment by Jarzka** [comment]
I have this similar problem while compiling my app with shadow-cljs: function parameters are undeclared. Works with lein-cljsbuild.

**Comment by roman01la** [comment]
@Jarzka I'd suggest opening a ticket in shadow-cljs then, feel free to refer to this issue

**Comment by Jarzka** [comment]
Actually when I compile with lein-cljsbuild, it builds the app but I get a lot of runtime errors like 

```
Uncaught Error: Doesn't support name: function (){
var ctr = (c.cljs$core$IFn$_invoke$arity$0 ? c.cljs$core$IFn$_invoke$arity$0() : c.call(null));
return ctr.apply(ctr,arguments);
}
    at Object.cljs$core$name [as name] (core.cljs:9524)
    at Object.daiquiri$util$fragment_QMARK_ [as fragment_QMARK_] (util.cljc:59)
    at Object.daiquiri$interpreter$interpret_vec [as interpret_vec] (interpreter.cljs:67)
    at Object.daiquiri$interpreter$interpret [as interpret] (interpreter.cljs:74)
    at Function.eval (main.cljs:337)
    at Function.eval [as cljs$core$IFn$_invoke$arity$2] (core.cljs:3912)
    at Function.eval [as cljs$core$IFn$_invoke$arity$2] (core.cljs:3948)
    at render (core.cljs:208)
    at eval (core.cljs:373)
    at eval (core.cljs:112)
```

However, I'm not yet sure if this is a bug in my code. This is my first time using rum so I need to investigate a bit further.

**Comment by roman01la** [comment]
@Jarzka could be something different, could you please post the code?

**Comment by Jarzka** [comment]
I'm implementing rum support for my library stylefy. The library contains examples on how to use rum and stylefy together. The example codebase is mostly a port from Reagent examples.

https://github.com/Jarzka/stylefy/tree/feature/rum/examples/rum

`lein cljsbuild once dev` -> Compiles, but gives runtime errors.
` shadow-cljs watch frontend` -> Unable to compile, gives undeclared var errors

**Comment by Jarzka** [comment]
I think I figured it out. Instead of this:

```cljs
(rum/defc main < rum/reactive []
  [top-level])
```


One should include sub-components like this:

```cljs
(rum/defc main < rum/reactive []
  (top-level))
```

With this, it works. I just need to fix all the sub-components. But I guess the shadow-cljs compile problem is indeed in shadow-cljs, not in rum.

I think it would be nice to have an example on how to use a sub-component in the Readme file. :)

**Comment by Jarzka** [comment]
I fixed some of my sub-component calls. Now I noticed that this sub-component:

```cljs
(rum/defc button < rum/reactive [text & [action-fn type]]
  [:div (use-style (button-style-by-type type)
                   {:on-click action-fn})
   text])
```

Gives the following error while running lein-cljsbuild:

```cljs
Uncaught TypeError: Cannot convert a Symbol value to a string
    at Array.join (<anonymous>)
    at Function.cljs$core$IFn$_invoke$arity$1 (core.cljs:2963)
    at Object.cljs$core$pr_writer_impl [as pr_writer_impl] (core.cljs:10174)
    at cljs$core$pr_writer (core.cljs:10183)
    at core.cljs:10306
    at Object.cljs$core$pr_sequential_writer [as pr_sequential_writer] (core.cljs:10037)
    at Object.cljs$core$print_prefix_map [as print_prefix_map] (core.cljs:10301)
    at Object.cljs$core$print_map [as print_map] (core.cljs:10315)
    at Object.cljs$core$pr_writer_impl [as pr_writer_impl] (core.cljs:10119)
    at Object.cljs$core$pr_writer [as pr_writer] (core.cljs:10183)
```

While using it as a sub component, like this:

```cljs
(rum/defc simple-examples < rum/reactive []
  [:div
   [:h1 "Generic button"]
   [:p "Just a simple styled button to begin with."]
   (button "Generic button")
   ...
   ]
```

Maybe there is some issue in rum after all?

EDIT: nevermind, this was also a bug in my own code 🙈

**Comment by Azzurite** [comment]
Yea I think the comments by @Jarzka are a different problem.

**Comment by stephenBDT** [comment]
The author of shadow-cljs was really helpful and figured that this must indeed be a rum issue:
https://github.com/thheller/shadow-cljs/issues/861

**Comment by thheller** [comment]
Reposting my answer here:

I think I found the cause.

The rum compiler macro uses a rather hacky way to disable warnings here

https://github.com/tonsky/rum/blob/599ef79c62da7af3fd27a2bddfc982b2815bc384/src/daiquiri/compiler.clj#L47-L49

but shadow-cljs explicitly checks whether a warning is disabled via false which the cljs.analyzer/no-warn macro would set. Maybe I should instead just treat the absense of a warning value as false instead of true. This is the first time I see this behavior.

So the code actually all compiles fine. The warning is just a side-effect from rum trying to infer the variable type/tag, which it intended to silence but did so improperly.

shadow-cljs defaults to `true` for warnings since it is the "better" default. Otherwise a macro may attempt to emit a custom warning during normal compilation which would then be immediately dropped since it defaulted to being ignored.

**Comment by Azzurite** [comment]
So the fix would just be to use `cljs.analyzer/no-warn` instead of `with-bindings* {(requiring-resolve 'cljs.analyzer/*cljs-warnings*) {}}`?

**Comment by thheller** [comment]
That would be the future proof way yes. Although the `requiring-resolve` trick won't work given that its a macro. Not sure why this is done in the first place though.

**Comment by roman01la** [comment]
Good catch, thanks. Lazy require was added to avoid hard dependency on ClojureScript for projects that are using Rum for templating on JVM, no cljs in dependencies. I don't have an answer for this issue right now.

**Comment by stephenBDT** [comment]
Maybe @tonsky can figure this out. I am new to clojure so I am not sure how to go forward from here?! 
Will this most likely also affect running with leinigen + figwheel?? (I hoped I could avoid those...)

For me it is imporant to be able to run rum on serverside and clientside.
Should I maybe just stick with the (very old) `[ua.modnakasta/rum "0.11.0-2"]` for now?

**Comment by roman01la** [comment]
@stephenBDT you can use a prev version of Rum, the most recent one doesn't include too many changes

**Comment by stephenBDT** [comment]
Okay, thank you @roman01la  :)
Sorry for being lazy, I don't want to go backwads and find the fist version where everything works. 
Do you accidentally know the latest version where this bug does not exist?

**Comment by roman01la** [comment]
a prev one

**Comment by Azzurite** [comment]
> I don't have an answer for this issue right now.

I have, though, maybe :) https://github.com/tonsky/rum/pull/234

**Comment by stephenBDT** [comment]
> a prev one

Sorry to bother you still, but a previous version does also not work for me!:
```
~/use-clj$ clj -M:run
Syntax error (FileNotFoundException) compiling at (daiquiri/compiler.clj:1:1).
Could not locate cljs/analyzer__init.class, cljs/analyzer.clj or cljs/analyzer.cljc on classpath.

Full report at:
/tmp/clojure-17750773598922002906.edn
```

Can be tested with this repo: https://github.com/stephenBDT/rum-seems-broken
I tried with `:exclusions [cljsjs/react cljsjs/react-dom]` and without....

Especially as a newcommer I am really struggling here.... 

I bet at least 80% wouldn't even have bothered so long with clojure and just moved on.
I know this is not anyones fault but I feel as library maintainer we should really make sure that the "official" tools (like `clj` with `deps.edn`) work out of the box... :(

**Comment by roman01la** [comment]
the last version fixes the issue in the prev version where the library didn't work w/o cljs dependency, thus you are having this error in the prev version of rum, it's not related

**Comment by roman01la** [comment]
resolved in 0.12.5, thanks @Azzurite for a quick fix
